### DISCUSSION TITLE
Spectral Peaks (Colorize waveforms based on sound brightness)

### DISCUSSION THREAD

**Discussion by X-Raym** [katspaugh/wavesurfer.js#2685] [Ideas]
This is just something to think about i don't think it's possible, or if it is it would probably be hard and take months to do.
Anyway, suggest it, in order to know what we can ask from an ultimate waveform display :P
Comparisonics is great because it lets us show the overall frequency of the sound on the waveform itself. It becomes indispensable for DJing and audio mixing analysis.
They do have an API http://www.comparisonics.com/api.html but i think it's written in C (no javascript ready =/)
However, Comparisonics have one major issue for me : the waveform is usualy (always ?) shown in mono, even in their own player and one website that integrated it (as Freesound) despite the fact they have stero files.

That's it for the idea, it far less important than having stereo waveform display, but I think it's pretty cool :P

Edit 2019: non need for comparisonics, Spectral Centroid algo can be used, see below.

**Comment by katspaugh** [comment]
Thanks for showing their site. There was demand for spectrograms. For instance, #74.
This might work, especially for simple sounds.

**Comment by X-Raym** [comment]
There is yest some differences between Comparisonics and Spectrogram. The first one use some advanced perceptive analysis and calculate the overgae over small slices of time, not per peak or per sample, and display the waveform with vertical stripes is far more human readable.
http://www.comparisonics.com/AES_ID31.pdf
It is far more easy to spot kick drum with Comparisonics :)

Spectrograms display frequency (from 20 to 20hz) (not amplitude) over time and look more like this http://en.wikipedia.org/wiki/File:Spectrogram-19thC.png wich can be good for spoting "noise" or "hum" frequencies.

**Comment by AnthonyDry** [comment]
This would be a sick improvement, but i guess it should be a plugin instead. I havent analysed the draw function at all but from the audiobuffer it shouldn't be impossible to analys  the audio thru bandpass filters and get peaks from general hz frequencies. such as 50hz and 5k hz for kickdrum.
Save the currentTime or(samplePosition) of the peaks to an array and then from the draw function set a different color. or make a vertical stripe based on the currentTimeArray or(sampleArray). :)

**Comment by X-Raym** [comment]
I made a bit of research and found that the algo behind that is called Spectral Centroid. Comparisonics used private patented algorithm but Spectral Centroid is a simple math funciton (free / open source etc).

You can see a demo of this implemented in JavaScript on waveforms on the nice [fav.js](https://github.com/flucoma/fav.js) library: 

![img](https://monosnap.com/image/ffurskujrDsIOjzQOzcq9jKzHMlZxY.png)

(fav.js has only few documentation but it works).

This feature is implemented in a large variety of software now (REAPER, Serato, RekordDJ, Samplitube, etc...) and since years on the FreeSound.org website. Really handy feature!

![on REAPER](https://monosnap.com/image/mSM6UJOu2i4gfrnrutC2uNwaORuMIA.png)

**Comment by katspaugh** [comment]
Looks damn cool!

**Comment by X-Raym** [comment]
@katspaugh any news about this?

**Comment by katspaugh** [comment]
I'll reopen it and move it to the Ideas forum.

**Comment by realies** [comment]
Any updates on this?

**Comment by johndpope** [comment]
you can use claude / grok to cut this - https://github.com/johndpope/comparisonics-hack/blob/main/test.py

https://grok.com/share/bGVnYWN5_b3340d59-14aa-4fc3-8b55-cdd229af0773

**Comment by SeanBannister** [comment]
Here's a spectral waveform demo with wavesurfer https://codepen.io/SeanBannister/full/PwGXapN click the "Browse Audio" button to load a file. It looks significantly more "defined" with heavily compressed music like EDM, for example this is an EDM track:

<img width="976" height="124" alt="image" src="https://github.com/user-attachments/assets/7b2a5cca-1ca4-4cbb-890f-203d82c38074" />

This is not:
<img width="967" height="113" alt="image" src="https://github.com/user-attachments/assets/2368a6b0-6038-42db-bf52-44a4896b44a0" />

I've coded all my life but there was no way I could of done this without Gemini 3.1 Pro. I'm shocked at how well it did. It did require a few hours and a lot of direction, I found this algorithm [github.com/turbo/libdjwaveform](https://github.com/turbo/libdjwaveform) which is written in C and now converted to JS. The code to actually color the waveform is only about 90 lines.

I added controls for "Low Crossover (Hz)" and "High Crossover (Hz)" so you can tweak the Red/Green/Blue frequencies and see them update on the waveform. I haven't had much time to tweak these, I think green needs to be more prominent, and maybe the color codes themselves need changing? I'd like to hear your feedback on sensible defaults.

At some point I would like to turn this into a wavesurfer plugin for easy implementation but I'll need to find the time.

**Comment by X-Raym** [comment]
@SeanBannister This is promising

For Reference, here how the waveform looks in REAPER (and the controls we have)

![Image](https://cloud.extremraym.com/sharex/2026-04/reaper_Ok5qSKAzwU.png)

https://cloud.extremraym.com/sharex/2026-04/F.gif

So it seems REAPER doesnt need adjustable band split and just use a hue shift (with adjustable hue start).
If I remember well, the algorithm in fav.js linked above was a bit closer.

**Reply by SeanBannister** [reply]
I did play around with fav.js, it certainly looks more defined and easier to read. The only reason I moved towards libdjwaveform is my experience with Traktor/VirtualDJ/Rekordbox DJ software, I'm so used to seeing those waveforms. And for my use case I'm actually displaying the waveforms in a stage lighting application where DJs can mark where lighting effects happen. When they do this they'll have both programs open at the same time so having the waveforms match makes it easier for them to read.

**Reply by X-Raym** [reply]
If what you done match existing software you use, it definitely makes sense indeed.
At least all this is a good proof of concept than some kind of frequency based colorization is possible, using frequency splitting (like your version) or not (like in REAPER / fav.js)

**Comment by SeanBannister** [comment]
I've turned my code into a wavesurfer plugin:
https://github.com/SeanBannister/wavesurfer.js/blob/master/src/plugins/spectral-peaks.ts

I also committed the ESM JS file so people can start testing it in their own projects:
https://github.com/SeanBannister/wavesurfer.js/blob/master/dist/plugins/spectral-peaks.esm.js

You can include it from CDN:
`import SpectralPeaks from "https://cdn.jsdelivr.net/gh/SeanBannister/wavesurfer.js@master/dist/plugins/spectral-peaks.esm.js";`

And I created a minimal code example to show how it works https://codepen.io/SeanBannister/pen/RNGEXjq

As for the code, I'm a bit worried about everything I'm doing in [renderWaveform](https://github.com/SeanBannister/wavesurfer.js/blob/c3f6dd2ba9590ccd326a60107fea5de3ae770646/src/plugins/spectral-peaks.ts#L145) and how that might conflict with other plugins. I just don't know enough about the codebase.
